src/github_runner_image_builder/cloud_image.py

```diff
@@ -0,0 +1,165 @@
+from github_runner_image_builder.utils import retry
+
+logger = logging.getLogger(__name__)
+
+SupportedBaseImageArch = typing.Literal["amd64", "arm64"]
+
+CHECKSUM_BUF_SIZE = 65536  # 65kb
+
+
+def download_and_validate_image(arch: Arch, base_image: BaseImage) -> Path:
+    """Download and verify the base image from cloud-images.ubuntu.com.
+
+    Args:
+        arch: The base image architecture to download.
+        base_image: The ubuntu base image OS to download.
+
+    Returns:
+        The downloaded image path.
+
+    Raises:
+        BaseImageDownloadError: If there was an error with downloading/verifying the image.
+    """
+    try:
+        bin_arch = _get_supported_runner_arch(arch)
+    except UnsupportedArchitectureError as exc:
+        raise BaseImageDownloadError from exc
+
+    image_path_str = f"{base_image.value}-server-cloudimg-{bin_arch}.img"
+    image_path = _download_base_image(
+        base_image=base_image, bin_arch=bin_arch, output_filename=image_path_str
+    )
+    shasums = _fetch_shasums(base_image=base_image)
+    if image_path_str not in shasums:
+        raise BaseImageDownloadError("Corresponding checksum not found.")
+    if not _validate_checksum(image_path, shasums[image_path_str]):
+        raise BaseImageDownloadError("Invalid checksum.")
+    return image_path
+
+
+def _get_supported_runner_arch(arch: Arch) -> SupportedBaseImageArch:
+    """Validate and return supported runner architecture.
+
+    The supported runner architecture takes in arch value from Github supported
+    architecture and outputs architectures supported by ubuntu cloud images.
+    See: https://docs.github.com/en/actions/hosting-your-own-runners/managing-\
+        self-hosted-runners/about-self-hosted-runners#architectures
+    and https://cloud-images.ubuntu.com/jammy/current/
+
+    Args:
+        arch: The compute architecture to check support for.
+
+    Raises:
+        UnsupportedArchitectureError: If an unsupported architecture was passed.
+
+    Returns:
+        The supported architecture.
+    """
+    match arch:
+        case Arch.X64:
+            return "amd64"
+        case Arch.ARM64:
+            return "arm64"
+        case _:
+            raise UnsupportedArchitectureError(f"Detected system arch: {arch} is unsupported.")
+
+
+@retry(tries=3, delay=5, max_delay=30, backoff=2, local_logger=logger)
+def _download_base_image(base_image: BaseImage, bin_arch: str, output_filename: str) -> Path:
+    """Download the base image.
+
+    Args:
+        bin_arch: The ubuntu cloud-image supported arch.
+        base_image: The ubuntu base image OS to download.
+        output_filename: The output filename of the downloaded image.
+
+    Raises:
+        BaseImageDownloadError: If there was an error downloaded from cloud-images.ubuntu.com
+
+    Returns:
+        The downloaded image path.
+    """
+    # The ubuntu-cloud-images is a trusted source
+    try:
+        urllib.request.urlretrieve(  # nosec: B310
```

Comment:
Can you set a timeout either by `socket.setdefaulttimeout(XX)` or using `request` instead?